src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala

```diff
@@ -0,0 +1,95 @@
+
+  def getGcsObjectRoleRequestUrl(agoraEntity: AgoraEntity, userEmail: String): String = {
+    val agoraBusiness = new AgoraBusiness
+    agoraBusiness.hasNamespaceNameId(agoraEntity) match {
+      case true =>
+        "https://www.googleapis.com/storage/v1/b/" + namespaceToBucketName(agoraEntity.namespace.get) + "/o/" + entityToObjectName(agoraEntity) + "/acl/user-" + userEmail
+      case false =>
+        throw new Exception("Trying to get Google ACLs for AgoraEntity with missing namespace, name, or snapshotId")
+    }
+  }
+
+  def getGcsBucketRoleRequestUrl(namespace: String, userEmail: String): String = {
+    "https://www.googleapis.com/storage/v1/b/" + namespaceToBucketName(namespace) + "/acl/user-" + userEmail
+  }
+
+  def insertGoogleBucketUrl(): String = {
+    "https://www.googleapis.com/storage/v1/b?" + AgoraConfig.gcsProjectId
+  }
+
+  def unrollGaclResponse(response:HttpResponse): GcsObjectRole = {
+    response.status match {
+      case OK =>
+        val GaclResponse = unmarshal[GoogleAccessControls].apply(response)
+        GcsObjectRole(GaclResponse.role.get) // String. READER or OWNER
+      case NotFound => GcsObjectRole(GcsRole.Nothing) // No ACL means the user has Nothing permission
+      case _ =>
+        // 500 and other errors.
+        logger.error("Error contacting GCS for object ACLs: " + response.status.toString() + response.entity.asString)
+        // Need to catch this error in the calling handler, so that requestContext can be completed
+        throw new ClientServiceFailure(response.status, response.entity.asString)
+    }
+  }
+
+  override def authorizationsForEntity(agoraEntity: Option[AgoraEntity], username: String): AuthorizedAgoraEntity = {
+    agoraEntity match {
+      case None =>  AuthorizedAgoraEntity(None, new AgoraPermissions(Nothing))
+      case Some(entity) =>
+        val responseFuture = pipeline(Get(getGcsObjectRoleRequestUrl(entity, username)))
+        val response = Await.result(responseFuture, Duration.Inf)
+        AuthorizedAgoraEntity(agoraEntity, RoleTranslator.gcsObjectToEntityPermissions(unrollGaclResponse(response)))
+    }
+  }
+
+  override def authorizationsForEntities(agoraEntities: Seq[AgoraEntity], username: String): Seq[AuthorizedAgoraEntity] = {
+    // Get a sequence of response Futures, one for each Some(entity). Package the http response alongside the entity.
+    // Transform the sequence of Futures into a Future of a sequence
+    val responseFutures: Future[Seq[GoogleResponse]] = Future.sequence(agoraEntities.map(entity => {
+      val httpResponseFuture = pipeline(Get(getGcsObjectRoleRequestUrl(entity, username)))
+      httpResponseFuture.map(httpResponse => new GoogleResponse(entity, httpResponse))
+    }))
+
+    // Futures can complete in parallel, and we await the result on the slowest HttpResponse
+    val googleResponses: Seq[GoogleResponse] = Await.result(responseFutures, Duration.Inf)
```

Comment:
Do we really want to wait forever or should we add a reasonable timeout? What happens if we get a timeout?